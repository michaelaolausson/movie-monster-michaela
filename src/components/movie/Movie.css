.media {
  background:#000000;
  height: 100%; 
}
.media-body {
  position: relative;
  background: #ffffff;
  color: #222222; 
  height: 100%;
}
.media-body p {
  background: #ffffff;
}
.media-body h3 {
  font-size: 1.3rem;
}
.add-to-list 
{
  font-size: 2em;
  position: absolute;
  top:0;
  right:0;
  color: #DDDDDD; 
}
.add-to-list .fa {
  margin-right: 0.25em;
  cursor: pointer;
}

