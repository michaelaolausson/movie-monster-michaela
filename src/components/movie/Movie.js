import React, { useState } from 'react';
import { Button, Col, Media, Alert } from 'react-bootstrap'
import './Movie.css';
import TMDB from '../../api/tmdb/TMDB'
const tmdb = new TMDB();

const Movie = (props) => {

  const [error, setError] = useState(null)

  const getImagePath = (path, width, height) => {
      let url = "https://image.tmdb.org/t/p/w" + width + "/" + path;
      if (!path) {
        url = "https://via.placeholder.com/" + width + "x" + height;
      } 
      return url;
  }

  const addToFavourites = async ( event ) => {
    try {
      await tmdb.manageFavourites(event.target.getAttribute('movieid'), true)
      setError(null)
    } catch (error) {
      setError("Couldn't add movie to favourites right now :(")
    } 
  }

  const removeFromFavourites = async ( event ) => {
    try {
      await tmdb.manageFavourites(event.target.getAttribute('movieid'), false) 
      setError(null)
    } catch {
      setError("Couldn't remove movie from favourites right now :(")
    }
  }
 
  const addToWatchlist = async ( event ) => {
    try {
      await tmdb.manageWatchlist(event.target.getAttribute('movieid'), true)
      setError(null)
    } catch {
      setError("Couldn't add movie to watchlist right now :(")
    }
  }

  const removeFromWatchlist = async ( event ) => {
    try {
      await tmdb.manageWatchlist(event.target.getAttribute('movieid'), false) 
      setError(null)
    } catch {
      setError("Couldn't remove movie from watchlist right now :(")
    }
  }

  if (!error) {
    return(
      <Col sm={12} md={6} xl={4} className="mb-4" key={props.i}>
          <Media>
              <img
                  width={"100px"}
                  src={getImagePath(props.imagePath, 200, 300)}
                  alt={ props.title + " - movie poster"}
              />
              <Media.Body className="p-3 d-flex flex-wrap">
                <div>
                  <h3>{props.title}</h3>
                  <p className="text-muted p-0 mb-2"><small>Release date: {props.year}</small></p>
                  <p className="p-0 mb-3">{props.description.substring(0,300) + "..."}</p>
                </div>
                  
                {props.favourite &&
                  <div className="align-self-end w-100">
                    <Button id="remove-from-favourites-btn" className="float-right" variant="danger"  movieid={props.id} onClick={removeFromFavourites}><i className="fa fa-heart-broken"></i> Remove from favourites</Button>
                  </div>
                }
  
                {props.watchlist &&
                  <div className="align-self-end w-100">
                    <Button id="remove-from-watchlist-btn" className="float-right" variant="danger"  movieid={props.id} onClick={removeFromWatchlist}><i className="fa fa-ban"></i> Remove from watchlist</Button>
                  </div>
                }
                
                {!props.favourite && !props.watchlist &&
                  <div className="align-self-end w-100">
                    <Button id="add-to-favourites-btn" className="float-right" variant="primary"  movieid={props.id} onClick={addToFavourites}><i className="fa fa-heart-o"></i> Add to favourites</Button>
                    <Button id="add-to-watchlist-btn" className="float-right" variant="secondary" movieid={props.id} onClick={addToWatchlist}><i className="fa fa-clock-o"></i> Add to watchlist</Button>
                  </div>
                }
              
              </Media.Body>
          </Media>
      </Col>
    )
  } else {
    return(
      <Col>
        <Alert id="error-box" variant="danger">{ error }</Alert>
      </Col>
    )
  }
}

export default Movie;
