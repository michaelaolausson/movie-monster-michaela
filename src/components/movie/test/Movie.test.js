import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { shallow } from "enzyme";
import Movie from "./../Movie";

describe("<Movie />", () => {
  const props = {
    id: 123,
    title: "Test",
    release_date: "1970-01-01",
    overview: "We donâ€™t serve their kind here! What? Your droids.",
    poster_path: "/2UDsmTvWSxVvuGoRQyV3zVaIRyu.jpg"
  };
  const wrapper = shallow(
    <Movie
      id={props.id}
      key={1}
      title={props.title}
      year={props.release_date}
      description={props.overview}
      imagePath={props.poster_path}
    />
  );
  it("should match the snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("throws no error when clicking on add favourite button", () => {
    const addFavouriteBtn = wrapper.find("#add-to-favourites-btn");
    addFavouriteBtn.simulate("click");
  });
  it("should render error element on error", () => {
    wrapper.error = "Something is wrong";
    wrapper.find("#error-box");
  });
  it("should not render error element when state error equals null", () => {
    wrapper.error = null;
    wrapper.find("#error-box");
  });
});
