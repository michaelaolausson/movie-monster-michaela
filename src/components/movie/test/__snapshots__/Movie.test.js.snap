// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Movie /> should match the snapshot 1`] = `"<div class=\\"mb-4 col-xl-4 col-md-6 col-sm-12\\"><div class=\\"media\\"><img width=\\"100px\\" src=\\"https://image.tmdb.org/t/p/w200//2UDsmTvWSxVvuGoRQyV3zVaIRyu.jpg\\" alt=\\"Test - movie poster\\"/><div class=\\"p-3 d-flex flex-wrap media-body\\"><div><h3>Test</h3><p class=\\"text-muted p-0 mb-2\\"><small>Release date: 1970-01-01</small></p><p class=\\"p-0 mb-3\\">We donâ€™t serve their kind here! What? Your droids....</p></div><div class=\\"align-self-end w-100\\"><button id=\\"add-to-favourites-btn\\" movieid=\\"123\\" type=\\"button\\" class=\\"float-right btn btn-primary\\"><i class=\\"fa fa-heart-o\\"></i> Add to favourites</button><button id=\\"add-to-watchlist-btn\\" movieid=\\"123\\" type=\\"button\\" class=\\"float-right btn btn-secondary\\"><i class=\\"fa fa-clock-o\\"></i> Add to watchlist</button></div></div></div></div>"`;
